oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,122 @@
+==============================
+
+.. list-table::
+
+   * - OEP
+     - :doc:`OEP-0031 <oep-0031-arch-i18n>`
+   * - Title
+     - Internationalization
+   * - Last Modified
+     - 2019-01-07
+   * - Authors
+     -
+   * - Arbiter
+     -
+   * - Status
+     - Draft
+   * - Type
+     - Architecture
+   * - Created
+     - 2019-01-07
+   * - Review Period
+     -
+
+.. contents::
+   :local:
+   :depth: 2
+
+Motivation
+==========
+
+Internationalization (i18n) is a vital part of unlocking content for users around the globe and effectively supporting our business partners. In order to help us reuse localized components from app to app, the different parts of the edX codebase should align on a consistent handling of i18n. This document covers our approach to internationalizing new front-end applications and components written in React.
+
+In this OEP we use the terms locale and language more or less interchangeably. However, note that this document addresses primarily making code multilingual, i.e., the issue of presenting the same content in different languages. Considering ways to make it multi-regional, to allow for presenting different content for different countries, is out of scope for this OEP.
+
+Technology Choice
+=================
+
+We will use `react-intl <https://github.com/yahoo/react-intl>`_ to internationalize React components, in agreement with `Paragon ADR-2 <https://github.com/edx/paragon/blob/f5b9f876800e5154ef01d04f682b8cda21b3e439/docs/decisions/0002-react-i18n.rst>`_. react-intl is in line with current industry standards in i18n and satisfies all currently identified i18n needs, including multiple plurals.
+
+Alternatives Considered
+-----------------------
+
+An alternative choice we considered was `i18next <https://react.i18next.com/>`_. **[TODO]**
+
+Locale
+======
+
+Localized content for new micro frontends should be at a URL that includes the language as the first part of the path after the domain. (Example: ``account.edx.org/es/``) The language string should be in the form of a `BCP 47 language tag <http://tools.ietf.org/html/rfc5646>`_. If the user enters the site at a URL that doesn't include a language string, we will determine a locale and then forward them to the corresponding URL. For consistency, this applies even if their language is English: ``account.edx.org/`` should be forwarded to ``account.edx.org/en/`` . We will use the most general, two-letter language codes to start with, and re-evaluate as necessary.
+
+When choosing what locale to use for a page, the locale in the URL always takes precedence. If there is none, check the user's cookies for a locale; if there isn't one, fetch it from the user's account settings and store it in their cookies for reuse. If there is no locale (or no locale we support) in the account settings, we will use the first locale we can support in the ``Accept-Language`` header that was sent with the request. Failing everything else, we will fall back to English.
```

Comment:
Is the plan to redirect users to the most appropriate URL path if not specified, or to render a language based on other heuristics at their un-locale-pathed URL?

Reply:
Yes, my intent in the paragraph above was to specify the "change the URL to what you think it should be" strategy.  Would it be clearer if that part of the earlier paragraph ("If the user enters the site...") came after this paragraph?

Reply:
It is not clear to me that this proposal recommends redirecting to a new URL based on this logic.  If we could make that more explicit that would be great.  It sounds like that is your intent and I agree with that as a proposal.